clang/test/CodeGenHLSL/builtins/dot.hlsl

```diff
@@ -0,0 +1,202 @@
+// RUN: %clang_cc1 -std=hlsl2021 -finclude-default-header -x hlsl -triple \
+// RUN:   dxil-pc-shadermodel6.3-library %s -fnative-half-type \
+// RUN:   -emit-llvm -disable-llvm-passes -O3 -o - | FileCheck %s
+// RUN: %clang_cc1 -std=hlsl2021 -finclude-default-header -x hlsl -triple \
+// RUN:   dxil-pc-shadermodel6.3-library %s -emit-llvm -disable-llvm-passes \
+// RUN:   -D__HLSL_ENABLE_16_BIT -o - | FileCheck %s --check-prefix=NO_HALF

```

Comment:
I am having a hard time seeing the difference between the `CHECK` and `NO_HALF` checks below. It looks like a lot of them are the same.


From the command line it looks like the difference is the `-fnative-half-type` flag which means that only the `half` tests should be different. I think in this case it is better to organize the check lines in a way that highlights those differences.

I would suggest using 3 check prefixes like `CHECK`, `NATIVE_HALF`, `NO_HALF`. The CHECK prefix will be common to both run lines and the two half ones will depend on the native-half-type flag.

```
RUN: clang ... -fnative-half-type ... | FileCheck %s -check-prefixes=CHECK,NATIVE_HALF
RUN: clang ...                    ... | FileCheck %s -check-prefixes=CHECK,NO_HALF
```

That way you only need to use the NATIVE_HALF and NO_HALF check prefixes when there should be actual differences between them.

Reply:
You are correct only lines 140-170 for the half cases are going to be different. They are supposed to be the same except for the float16 cases. I'm not sure I agree with the suggestion, `NO_HALF` is how we are doing all the other intrinsics I don't understand why this needs to be a special case..,

That said I can model these tests more like sin.hlsl where we only do the half checks on the half code:
https://github.com/llvm/llvm-project/blob/main/clang/test/CodeGenHLSL/builtins/sin.hlsl#L8C1-L35C2

And not like https://github.com/llvm/llvm-project/blob/main/clang/test/CodeGenHLSL/builtins/max.hlsl where we do it on every test case if you think that would add some clarity.

Reply:
This Pr fixed a bunch of the  issues mentioned: https://github.com/llvm/llvm-project/pull/81782/files
I've updated these tests to be similar.

Reply:
The test looks much better now and makes a lot more sense to me.

I still think we are missing out on opportunities to strengthen our testing. The two RUN lines are checking the cases where we have native 16-bit types or not. There are many CHECKS here that are only being validated when 16-bit types are enabled, but are fully valid when 16-bit types are not enabled.

For example,
```
// CHECK: %dx.dot = mul i32 %0, %1
// CHECK: ret i32 %dx.dot
int test_dot_int ( int p0, int p1 ) {
  return dot ( p0, p1 );
}
```
is equally valid whether 16-bit types are enabled or not, but we only check it when 16-bit types are enabled. 

If we used 3 different check prefixes (`CHECK`, `NO_HALF`, `NATIVE_HALF`) then it would give better coverage and make it clearer where we expect the differences to be.